File: netDxf/Entities/Polyline3D.Topology.cs
using System;

namespace netDxf.Entities
{
    public partial class Polyline3D
    {
        /// <summary>Inserts a point and, for a retained sequence, a new owned VERTEX identity.</summary>
        /// <param name="index">The insertion index, from zero through Vertexes.Count.</param>
        /// <param name="position">The finite position of the inserted point.</param>
        /// <remarks>
        /// Retained sequences must be registered in their source document and keep their source
        /// DXF profile. The new record uses the current parent layer and ordinary POLYLINE ownership;
        /// it does not copy optional fields or metadata from another vertex. Unaffected records and
        /// SEQEND keep their identities. No handle is allocated when validation rejects the edit.
        /// </remarks>
        public void InsertVertex(int index, Vector3 position)
        {
            if (index < 0 || index > this.vertexes.Count) throw new ArgumentOutOfRangeException(nameof(index));
            ValidateTopologyPosition(position, nameof(position));
            if (this.vertexes.Count >= 65536) throw new NotSupportedException("A polyline topology edit cannot exceed 65,536 vertices.");
            DxfDocument document = this.ValidateTopologyEdit();
            if (!this.HasStoredRecords) { this.vertexes.Insert(index, position); return; }
            Polyline3DRecord inserted = document.PreparePolylineVertexInsertion(this, position);
            // Reserve list storage before committing registration. The new record has no user
            // metadata or callbacks; registration only installs internal metadata bookkeeping.
            if (this.vertexes.Capacity < this.vertexes.Count + 1) this.vertexes.Capacity = this.vertexes.Count + 1;
            if (this.storedVertexRecords.Capacity < this.storedVertexRecords.Count + 1) this.storedVertexRecords.Capacity = this.storedVertexRecords.Count + 1;
            document.RegisterPolylineVertexInsertion(inserted);
            this.vertexes.Insert(index, position);
            this.storedVertexRecords.Insert(index, inserted);
        }

        /// <summary>Removes a point and its retained VERTEX after checking incoming dependencies.</summary>
        /// <param name="index">The zero-based index to remove.</param>
        /// <remarks>
        /// Retained sequences must be registered in their source document. Incoming semantic
        /// references and private or owned record payloads reject before mutation. Removal does
        /// not cascade to referenced objects. A removed record retains its retired handle for
        /// inspection, has no owner and cannot be inserted or adopted again. A retained sequence
        /// must keep at least two points, as required by the existing writer. SEQEND stays stable.
        /// </remarks>
        public void RemoveVertexAt(int index)
        {
            if (index < 0 || index >= this.vertexes.Count) throw new ArgumentOutOfRangeException(nameof(index));
            DxfDocument document = this.ValidateTopologyEdit();
            if (!this.HasStoredRecords) { this.vertexes.RemoveAt(index); return; }
            if (this.vertexes.Count <= 2) throw new InvalidOperationException("A retained 3D polyline must keep at least two vertices for output.");
            Polyline3DRecord removed = this.storedVertexRecords[index];
            document.ValidatePolylineVertexRemoval(removed);
            document.UnregisterPolylineVertexRemoval(removed);
            this.vertexes.RemoveAt(index);
            this.storedVertexRecords.RemoveAt(index);
            removed.Owner = null;
            removed.IsRemoved = true;
        }

        /// <summary>Moves a point and its retained VERTEX identity to a final zero-based index.</summary>
        /// <param name="fromIndex">The current zero-based index.</param>
        /// <param name="toIndex">The final zero-based index after moving the point.</param>
        /// <remarks>
        /// Both indices must identify existing points. Metadata follows the moved identity and
        /// no handles are allocated. Equal indices still validate the current sequence. Retained
        /// sequences must be registered in their source document and keep their source profile.
        /// </remarks>
        public void MoveVertex(int fromIndex, int toIndex)
        {
            if (fromIndex < 0 || fromIndex >= this.vertexes.Count) throw new ArgumentOutOfRangeException(nameof(fromIndex));
            if (toIndex < 0 || toIndex >= this.vertexes.Count) throw new ArgumentOutOfRangeException(nameof(toIndex));
            this.ValidateTopologyEdit();
            if (fromIndex == toIndex) return;
            Vector3 point = this.vertexes[fromIndex];
            this.vertexes.RemoveAt(fromIndex);
            this.vertexes.Insert(toIndex, point);
            if (this.HasStoredRecords)
            {
                Polyline3DRecord record = this.storedVertexRecords[fromIndex];
                this.storedVertexRecords.RemoveAt(fromIndex);
                this.storedVertexRecords.Insert(toIndex, record);
            }
        }

        private DxfDocument ValidateTopologyEdit()
        {
            if (this.vertexes.Count > 65536) throw new NotSupportedException("A polyline topology edit cannot exceed 65,536 vertices.");
            foreach (Vector3 position in this.vertexes) ValidateTopologyPosition(position, "Vertexes");
            if (!this.HasStoredRecords) return null;
            DxfDocument document = this.storedRecordDocument;
            if (document == null || this.Handle == null || !ReferenceEquals(document.GetObjectByHandle(this.Handle), this)
                || this.Owner == null || !ReferenceEquals(document.GetObjectByHandle(this.Owner.Handle), this.Owner))
                throw new InvalidOperationException("Retained polyline topology edits require a registered parent in its source document; adopt a qualified clone before editing.");
            this.ValidateStoredRecords(document, true);
            return document;
        }

        private static void ValidateTopologyPosition(Vector3 position, string parameter)
        {
            if (double.IsNaN(position.X) || double.IsInfinity(position.X)
                || double.IsNaN(position.Y) || double.IsInfinity(position.Y)
                || double.IsNaN(position.Z) || double.IsInfinity(position.Z))
                throw new ArgumentException("Polyline topology coordinates must be finite.", parameter);
        }
    }
}

File: javascript/netDxf/Entities/Polyline3D.Topology.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { Stored3D } from './Polyline3D.StoredRecords.js';
import { ArgumentException, InvalidOperationException, NotSupportedException, RequireInteger } from '../../runtime/Errors.js';
function finite(p,name){if(![p.X,p.Y,p.Z].every(Number.isFinite))throw new ArgumentException('Topology coordinates must be finite.',name);}
function validate(p){
  if(p.Vertexes.Count>65536)throw new NotSupportedException('A topology edit cannot exceed 65,536 vertices.');
  for(const v of p.Vertexes)finite(v,'Vertexes');if(!p.HasStoredRecords)return null;
  const doc=Stored3D(p).document;
  if(doc===null||p.Handle===null||doc.GetObjectByHandle(p.Handle)!==p||p.Owner===null||doc.GetObjectByHandle(p.Owner.Handle)!==p.Owner)throw new InvalidOperationException('Retained topology requires a registered source parent.');
  p.ValidateStoredRecords(doc,true);return doc;
}
export function InstallPolyline3DTopology(Type){
  Type.prototype.InsertVertex=function(index,position){
    RequireInteger(index,0,this.Vertexes.Count,'index');finite(position,'position');if(this.Vertexes.Count>=65536)throw new NotSupportedException('A topology edit cannot exceed 65,536 vertices.');
    const doc=validate(this);if(!this.HasStoredRecords){this.Vertexes.Insert(index,position);return;}
    const inserted=doc.PreparePolylineVertexInsertion(this,position);doc.RegisterPolylineVertexInsertion(inserted);this.Vertexes.Insert(index,position);Stored3D(this).vertices.Insert(index,inserted);
  };
  Type.prototype.RemoveVertexAt=function(index){
    RequireInteger(index,0,this.Vertexes.Count-1,'index');const doc=validate(this);if(!this.HasStoredRecords){this.Vertexes.RemoveAt(index);return;}
    if(this.Vertexes.Count<=2)throw new InvalidOperationException('A retained polyline must keep at least two vertices.');const r=Stored3D(this).vertices.get_Item(index);
    doc.ValidatePolylineVertexRemoval(r);doc.UnregisterPolylineVertexRemoval(r);this.Vertexes.RemoveAt(index);Stored3D(this).vertices.RemoveAt(index);r.Owner=null;r.IsRemoved=true;
  };
  Type.prototype.MoveVertex=function(fromIndex,toIndex){
    RequireInteger(fromIndex,0,this.Vertexes.Count-1,'fromIndex');RequireInteger(toIndex,0,this.Vertexes.Count-1,'toIndex');validate(this);if(fromIndex===toIndex)return;
    const p=this.Vertexes.get_Item(fromIndex);this.Vertexes.RemoveAt(fromIndex);this.Vertexes.Insert(toIndex,p);
    if(this.HasStoredRecords){const a=Stored3D(this).vertices;const r=a.get_Item(fromIndex);a.RemoveAt(fromIndex);a.Insert(toIndex,r);}
  };
}
